pkg/ext/stores/tokens/tokens.go

```diff
@@ -0,0 +1,922 @@
+
+	token, err := t.SystemStore.Get(name, opts)
+	if err != nil {
+		return nil, err
+	}
+	if _, err := t.checker.UserHasPermission("get", token, ctx); err != nil {
+		return nil, err
+	}
+
+	return token, nil
+}
+
+func (t *SystemStore) Get(name string, opts *metav1.GetOptions) (*ext.Token, error) {
+	// Core token retrieval from backing secrets
+	currentSecret, err := t.secretClient.Get(TokenNamespace, name, *opts)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, err
+		}
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to retrieve token %s: %w", name, err))
+	}
+	token, err := tokenFromSecret(currentSecret)
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to extract token %s: %w", name, err))
+	}
+	token.Status.TokenValue = ""
+	return token, nil
+}
+
+func (t *Store) List(ctx extcore.Context, opts *metav1.ListOptions) (*ext.TokenList, error) {
+	// cannot use checkForManageToken here. we have lots of tokens to check, with the same admin value.
+	hasPermission, err := t.checker.UserHasManageTokenPermissions(ctx)
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to check user permissions: %w", err))
+	}
+	return t.SystemStore.list(hasPermission, ctx.User.GetName(), opts)
+}
+
+func (t *SystemStore) List(opts *metav1.ListOptions) (*ext.TokenList, error) {
+	return t.list(true, "", opts)
+}
+
+func (t *SystemStore) list(hasPermission bool, user string, opts *metav1.ListOptions) (*ext.TokenList, error) {
+	// Core token listing from backing secrets
+	secrets, err := t.secretClient.List(TokenNamespace, *opts)
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to list tokens: %w", err))
+	}
+	var tokens []ext.Token
+	for _, secret := range secrets.Items {
+		token, err := tokenFromSecret(&secret)
+		if err != nil {
+			// ignore tokens with broken information
+			continue
+		}
+		// users can only list their own tokens, unless they have full permissions on this group
+		if !hasPermission && token.Spec.UserID != user {
+			continue
+		}
+		tokens = append(tokens, *token)
+	}
+	list := ext.TokenList{
+		ListMeta: metav1.ListMeta{
+			ResourceVersion: secrets.ResourceVersion,
+		},
+		Items: tokens,
+	}
+	return &list, nil
+}
+
+func (t *Store) Watch(ctx extcore.Context, opts *metav1.ListOptions) (<-chan extcore.WatchEvent[*ext.Token], error) {
+	// cannot use checkForManageToken here. we have lots of tokens to check, with the same admin value.
+	hasPermission, err := t.checker.UserHasManageTokenPermissions(ctx)
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to check user permissions: %w", err))
+	}
+
+	ch := make(chan extcore.WatchEvent[*ext.Token])
+
+	go func() {
+		// watch backend secret, transform its events into token events
+		watcher, err := t.secretClient.Watch(TokenNamespace, metav1.ListOptions{
+			ResourceVersion: opts.ResourceVersion,
```

Comment:
I'm pretty sure all list options `opts` have to merged or copied and reused here.
A watch can be called with different options (e.g. field or label selectors) and they have to be respected.